test/Mono.Linker.Tests.Cases/Reflection/ObjectGetType.cs

```diff
@@ -4,61 +4,792 @@
+				// that it reuses the already processed info correctly.
+				GetInstance ().GetType ().GetMethod ("InterfaceMethod");
+			}
+		}
+
+		[Kept]
+		class AnnotationsRequestedOnImplementation
+		{
+			[Kept]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)]
+			interface IAnnotatedInterface
+			{
+				[Kept] // Kept because it's implemented on the class
+				void InterfaceMethod ();
+
+				// Annotation will not be applied to the interface, since nothing
+				// asked for it via reflection.
+				static void DoSomething () { }
+
+				void DefaultInterfaceMethod () { }
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptInterface (typeof (IAnnotatedInterface))]
+			class ImplementationClass : IAnnotatedInterface
+			{
+				// Annotation will be applied to the implementation type which the reflection
+				// asked for
+				[Kept]
+				void IAnnotatedInterface.InterfaceMethod ()
+				{
+				}
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptBaseType (typeof (ImplementationClass))]
+			class Derived : ImplementationClass
+			{
+				// Annotation will be applied to a derived type as well
+				[Kept]
+				public void NewMethod () { }
+			}
+
+			[Kept]
+			static ImplementationClass GetInstance () => new Derived ();
+
+			[Kept]
+			public static void Test ()
+			{
+				// Make sure the interface is kept
+				var i = typeof (IAnnotatedInterface);
+				GetInstance ().GetType ().GetMethod ("InterfaceMethod");
+			}
+		}
+
+		[Kept]
+		class AnnotationsRequestedOnInterface
+		{
+			[Kept]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods | DynamicallyAccessedMemberTypes.NonPublicMethods)]
+			interface IAnnotatedInterface
+			{
+				[Kept] // Kept because it's implemented on the class
+				void InterfaceMethod ();
+
+				// Annotation applied to the interface since that's what reflection asked about
+				[Kept]
+				static void DoSomething () { }
+
+				[Kept]
+				void DefaultInterfaceMethod () { }
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptInterface (typeof (IAnnotatedInterface))]
+			class ImplementationClass : IAnnotatedInterface
+			{
+				[Kept]
+				void IAnnotatedInterface.InterfaceMethod ()
+				{
+				}
+			}
+
+			[Kept]
+			[KeptMember (".ctor()")]
+			[KeptBaseType (typeof (ImplementationClass))]
+			class Derived : ImplementationClass
+			{
+				[Kept]
+				public void NewMethod () { }
+			}
+
+			[Kept]
+			static IAnnotatedInterface GetInstance () => new Derived ();
+
+			[Kept]
+			public static void Test ()
+			{
+				// Make sure the interface is kept
+				var i = typeof (IAnnotatedInterface);
+				GetInstance ().GetType ().GetMethod ("InterfaceMethod");
+			}
+		}
+
+		[Kept]
+		class AllAnnotationsAreApplied
```

Comment:
How well does the code handle complex or conflicting hierarchies? I couldn't find any test and it could be worth adding as well.

Examples like

```c#
[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.All)]
class X : A, I
{
}

[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.None)]
class A : I
{
	public virtual void M ();
}

[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.None)]
interface I
{
	void M ();
}
```

or 

```c#
[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.None)]
class C : B
{
	public override  void Foo () {}
}

[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.All)]
class B : A
{
	public override  void Foo () {}
}

[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.None)]
class A
{
	public virtual void Foo () {}
}
```

Reply:
Good suggestions - I added a bit more complicated case of both samples above and one more around diamond shapes with interfaces.